# Base scRNAseq analysis Snakemake that combines all other Snakefiles in ./submodules together. 

##### Load Config and Sample Sheets #####
import easygui
print("Please choose a configuration file to run Snakemake with.")
config_file = easygui.fileopenbox()

# Include common settings
include: "./submodules/common.snakefile"

# Modulization through Submodules: 
subworkflow annotate_raw_sequences:
    workdir:
        "./submodules/annotate_raw_sequences"
    snakefile:
        "./submodules/annotate_raw_sequences/Snakefile"        
subworkflow star_index:
    workdir:
        "./submodules/star_index"
    snakefile:
        "./submodules/star_index/Snakefile"
subworkflow star_align:
    workdir:
        "./submodules/star_align"
    snakefile:
        "./submodules/star_align/Snakefile"
subworkflow map_barcodes:
    workdir:
        "./submodules/map_barcodes"
    snakefile:
        "./submodules/map_barcodes/Snakefile"
subworkflow generate_counttable:
       workdir:
        "./submodules/generate_counttable"
    snakefile:
       "./submodules/generate_counttable/Snakefile"

# Main workflow rule, determining output of workflow
rule all:
    input:
        annotate_raw_sequences( expand("{indexdir}/{indexfile}"                                     , indexdir=indexdir, indexfile=indexfiles)),
        star_index(             expand("{indexdir}/{indexfile}"                                     , indexdir=indexdir, indexfile=indexfiles)),
        star_align(             expand('{tmpstore}/star_align/{fileprefix}/Aligned.out.bam'        , tmpstore=tmpstore, fileprefix=fileprefixes)),
        map_barcodes(           expand('{tmpstore}/mapping/{fileprefix}/counts.tsv'                 , tmpstore=tmpstore, fileprefix=fileprefixes)),
        generate_counttable(    expand('{output}/generate_counttable/counttables_combined_ALL.csv'  , output=output)),



