# Base scRNAseq analysis Snakemake that combines all other Snakefiles in ./submodules together. 

# Load config
import easygui
print("Please choose a configuration file to run Snakemake with.")
config_file = easygui.fileopenbox()
#config_file = 'snakemake_config.yaml'

# Include common settings
include: "./common.snakefile"

# Modulization through Submodules:     
subworkflow star_index:
    workdir:
        "./submodules/star_index"
    snakefile:
        "./submodules/star_index/Snakefile"
subworkflow star_align:
    workdir:
        "./submodules/star_align"
    snakefile:
        "./submodules/star_align/Snakefile"
subworkflow map_barcodes:
    workdir:
        "./submodules/map_barcodes"
    snakefile:
        "./submodules/map_barcodes/Snakefile"
subworkflow generate_counttable:
    workdir:
        "./submodules/generate_counttable"
    snakefile:
       "./submodules/generate_counttable/Snakefile"

# Main workflow rule, determining output of workflow
rule all:
    input:
        star_index(         expand("{indexdir}/{reference}/{indexfiles}", indexdir = config['indexdir'], reference = config['reference'], indexfiles = indexfiles )),
        star_align(         expand('{tmppath}/star_align/{reference}/{fileprefixes}/Aligned.out.bam', tmppath = tmppath, reference = config['reference'], fileprefixes= fileprefixes )),
        map_barcodes(       expand('{tmppath}/mapping/{reference}/{fileprefixes}/counts.tsv', tmppath = tmppath, reference = config['reference'], fileprefixes= fileprefixes )),
        generate_counttable(expand('{outputpath}/count_tables/ALL.csv', outputpath = outputpath ))
    threads: 8


