---
output: html_document
params:
  featuretable: /data/mouse-lymph-node-stromal-cell-time-series/snakemake_temporary_files/mapping/mouse-grcm38/LNS2_1/HqFeatureCounts.tsv
  cellbcfile: /data/mouse-lymph-node-stromal-cell-time-series/annotations_common/celseq_barcodes.192.txt
  countsfile: /data/mouse-lymph-node-stromal-cell-time-series/snakemake_temporary_files/mapping/mouse-grcm38/LNS2_1/counts.tsv
  samplename: LNS_W0_P4_3
  UMIcardinality: 256
  UMIcorrection: TRUE
---

---
title: Analysis of sample `r params$samplename`
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
```

```{r parameters}
# Note that these parameter settings are overriden in the rendering R-script 
# (render_reports.R), not in the Snakefile
mapfile <- params[['featuretable']]
cellbcfile <- params[['cellbcfile']]
countsfile <- params[['countsfile']]
UMIcardinality <- as.integer(params[['UMIcardinality']])
UMIcorrection <- as.logical(params[['UMIcorrection']])

# Settings for testing this script directly:
# mapfile <- '/data/mouse-lymph-node-stromal-cell-time-series/snakemake_temporary_files/mapping/mouse-grcm38/LNS2_1/HqFeatureCounts.tsv'
# cellbcfile <- '/data/mouse-lymph-node-stromal-cell-time-series/annotations_common/celseq_barcodes.192.txt'
# countsfile <- '/data/mouse-lymph-node-stromal-cell-time-series/snakemake_temporary_files/mapping/mouse-grcm38/LNS2_1/counts.tsv'
# UMIcardinality <- as.integer('256')
# UMIcorrection <- as.logical('TRUE')

d <- read_tsv(mapfile, comment='@', na='NA')
bc <- read_tsv(cellbcfile, col_names = c('barcodenr','cellbc'))
```

```{r counts}
counts_per_cell_and_gene <- bc %>%
  left_join(d) %>%
  distinct(cellbc, geneid, umi) %>%
  group_by(cellbc, geneid) %>%
  summarise(reads = n())

# To infer the number of transcripts of a gene, a correction on the number of 
# observed UMI's per gene can be made that accounts for the possibility that the
# same UMI is attached to different RNA molecules of the same gene. See eq. 3 in
# Grun et al. (2014), Nature Methods 11:637--640 
# http://dx.doi.org/10.1038/nmeth.2930
if (UMIcorrection) {
    if (UMIcardinality < 4) {
        # TODO: error message, UMI cardinality rediculously low, do nothing
    } else {
        counts_per_cell_and_gene <- counts_per_cell_and_gene %>%
            mutate(transcripts = -UMIcardinality * log(1 - (reads/UMIcardinality)))
    }
}

# write_tsv(counts_per_cell_and_gene %>% filter(!is.na(geneid)), path=countsfile)
write_tsv(counts_per_cell_and_gene, path=countsfile)

```

```{r readdist}
reads_per_cell <- counts_per_cell_and_gene %>%
  ungroup() %>%
  mutate(matchedGene = ifelse(!is.na(geneid), 'matched', 'unmatched')) %>%
  group_by(cellbc, matchedGene) %>%
  summarise(totalreads=sum(reads)) %>%
  ungroup() %>%
  complete(cellbc, matchedGene, fill=list(totalreads=0))

reads_per_cell$matchedGene <- factor(reads_per_cell$matchedGene) %>% relevel('unmatched')

orderseq <- reads_per_cell %>% 
  filter(matchedGene=='matched') %>%
  arrange(desc(totalreads)) %>%
  mutate(sortorder = row_number()) %>% 
  select(cellbc, sortorder)

reads_per_cell <- reads_per_cell %>% 
  inner_join(orderseq)
```

Figure 1 shows the number of unique reads for each of the `r length(bc$barcodenr)` cells in the sample, either matching or not matching with a mouse gene. The cells were ordered by decreasing number of unique gene-matching reads.  

```{r, fig.cap='Fig 1. The number of unique reads matching or not matching a gene.'}
ggplot(reads_per_cell, aes(x=sortorder, y=totalreads, fill=matchedGene)) +
  geom_area() + 
  labs(x='Cell number', y='Mapped unique reads', fill='Matching') 
```

```{r, fig.cap='Fig 2. Histogram of number of expressed genes per cell'}
expressed_genes_per_cell <- counts_per_cell_and_gene %>%
  distinct(cellbc, geneid) %>%
  group_by(cellbc) %>%
  summarise(nr_of_genes = n())

ggplot(expressed_genes_per_cell, aes(x=nr_of_genes)) + 
  geom_histogram() +
  labs(x='Number of expressed genes', y='Number of cells')
```

```{r}
mtreads <- bc %>% 
  inner_join(d) %>%
  distinct(cellbc, geneid, reference, umi) %>%
  mutate(origin = ifelse(reference=='MT', 'mt', 'cell')) %>%
  group_by(cellbc, origin) %>%
  summarise(reads = n()) %>%
  spread(origin, reads, 0)
```


```{r, fig.cap='Number of expressed genes versus number of unique reads in cells'}
reads_vs_genes <- reads_per_cell %>%
  filter(matchedGene=='matched') %>%
  inner_join(expressed_genes_per_cell)

ggplot(reads_vs_genes, aes(x=totalreads, y=nr_of_genes)) +
  geom_point() + 
  labs(x='Number of unique reads', y='Number of expressed genes')
```


<!-- # number of reads mapping to a gene but missed because of invalid cell barcode -->
<!-- # reads_per_cat <- d %>% -->
<!-- #   filter(!is.na(geneid)) %>% -->
<!-- #   left_join(bc) %>% -->
<!-- #   mutate(hasCellBC = ifelse(!is.na(barcodenr), 'yes', 'no')) %>% -->
<!-- #   group_by(hasCellBC) %>% -->
<!-- #   summarise(totalReads = n()) -->
<!-- #    -->
<!-- # ``` -->
<!-- # ```{r} -->
<!-- # disttable <- counts_per_cell_and_gene %>%  -->
<!-- #   spread(cellbc, reads, 0) -->
<!-- #  -->
