# snakemake workflow for generating STAR index files of a reference sequence

include: "../common.snakefile"

# STAR extra options
extra = ' '.join(config['params']['starindex'][config['annotationformat']])

# Expected index files
indexfiles = ['SA','SAindex','chrLength.txt','chrName.txt','chrNameLength.txt']

##### target rules #####
# pseudorule consuming all the final output
rule all:
    input: expand("{indexdir}/{indexfile}", indexdir=indexdir, indexfile=indexfiles)

# unpacking the compressed fasta file
# rule decompress:
#     input: compressedfastafile
#     output: fastafile
#     shell: 
#         'unpigz -k {compressedfastafile}' 

# the alignment by STAR
rule index:
    input:
        fastafile = fastafile,
        gtffile = gtffile
    output:
        expand("{indexdir}/{indexfile}", indexdir=indexdir, indexfile=indexfiles)
    shell:
        'STAR --runThreadN {threads} --runMode genomeGenerate --genomeDir {indexdir} --genomeFastaFiles {input[fastafile]} --sjdbGTFfile {input[gtffile]} {extra}'
