# Snakemake Workflow for Module 1: Generating STAR Index Files of a reference sequence

# Include common settings
include: "../common.snakefile"

# STAR extra options
extra = ' '.join(config['params']['starindex'][config['annotationformat']])

##### Target Rules #####

# Pseudorule consuming all the final output
rule all:
    input: expand("{indexdir}/{indexfile}", indexdir = indexdir, indexfile = indexfiles)

# Unpacking the compressed fasta file
rule decompress:
    input: 
        compressedfastafile
    output: 
        fastafile = str(data / config['refdir'] / config['reference'] / (references.loc[config['reference'], 'genomefile'])).replace(".gz","")
    shell: 
        'unpigz -k {compressedfastafile}'

# Rule for Module 2: The alignment by STAR
rule index:
    input:
        fastafile = rules.decompress.output,
        gtffile = gtffile
    output:
        expand("{indexdir}/{indexfile}", indexdir = indexdir, indexfile = indexfiles)
    threads:
        24
    shell:
        'STAR --runThreadN {threads} --runMode genomeGenerate --genomeDir {indexdir} --genomeFastaFiles {input[fastafile]} --sjdbGTFfile {input[gtffile]} {extra}'


