# Snakemake Workflow for Module 1: Generating STAR Index Files of a reference sequence

# Include common settings
include: "../common.snakefile"

# STAR extra options
extra = ' '.join(config['params']['starindex'][config['annotationformat']])

##### Target Rules #####

# Pseudorule consuming all the final output
rule all:
    input: expand("{indexdir}/{indexfile}", indexdir=indexdir, indexfile=indexfiles)

# Rule for Module 2: The alignment by STAR
rule index:
    input:
        fastafile = fastafile,
        gtffile = gtffile
    output:
        expand("{indexdir}/{indexfile}", indexdir=indexdir, indexfile=indexfiles)
    shell:
        'STAR --runThreadN {threads} --runMode genomeGenerate --genomeDir {indexdir} --genomeFastaFiles {input[fastafile]} --sjdbGTFfile {input[gtffile]} {extra}'
